mod-infra/src/main/scala/at/forsyte/apalache/infra/passes/options.scala

```diff
@@ -234,3 +249,298 @@ object SMTEncoding {
+      config <- checker.config.toTry("checker.config")
+      discardDisabled <- checker.discardDisabled.toTry("checker.discardDisabled")
+      init <- checker.init.toTry("checker.init")
+      inv <- checker.inv.toTry("checker.inv")
+      length <- checker.length.toTry("checker.length")
+      maxError <- checker.maxError.toTry("checker.maxError")
+      next <- checker.next.toTry("checker.next")
+      noDeadlocks <- checker.noDeadlocks.toTry("checker.noDeadlocks")
+      nworkers <- checker.nworkers.toTry("checker.nworkers")
+      smtEncoding <- checker.smtEncoding.toTry("checker.smtEncoding")
+      temporal <- checker.temporal.toTry("checker.temporal")
+      tuning <- checker.tuning.toTry("checker.tuning")
+      view <- checker.view.toTry("checker.view")
+    } yield Checker(
+        algo = algo,
+        cinit = cinit,
+        config = config,
+        discardDisabled = discardDisabled,
+        init = init,
+        inv = inv,
+        length = length,
+        maxError = maxError,
+        next = next,
+        noDeadlocks = noDeadlocks,
+        nworkers = nworkers,
+        smtEncoding = smtEncoding,
+        temporal = temporal,
+        tuning = tuning,
+        view = view,
+    )
+  }
+
+  /**
+   * Provides a configured [[OptionGroup]] instance via a Google Guice `Provider`
+   *
+   * The provider enables late-binding of a configured option instance, which is then made available to any inject
+   * annotated class classes via google Guice's dependency injection.
+   *
+   * For more on Guice providers, see https://github.com/google/guice/wiki/InjectingProviders
+   */
+  class Provider[O <: OptionGroup] extends inject.Provider[O] {
+    protected var _options: Option[O] = None
+
+    /**
+     * Set the option to be provided
+     *
+     * This is typically called once, after parsing [[Config]]s into [[OptionGroup]]s. Use of it is to mutate the stored
+     * options to communicate between passes is strongly discouraged (but not forbidden).
+     */
+    def set(options: O): Unit = _options = Some(options)
+
+    /** Get the provided option */
+    def get(): O = _options.getOrElse(throw new inject.ProvisionException("pass options were not configured"))
+  }
+
+  ////////////////
+  // Interfaces //
+  ////////////////
+
+  // The following traits specify combinations of needed options.
+
+  trait HasCommon extends OptionGroup {
+    val common: Common
+  }
+
+  trait HasInput extends HasCommon {
+    val input: Input
+  }
+
+  trait HasOutput extends HasCommon {
+    val output: Output
+  }
+
+  trait HasIO extends HasInput with HasOutput
+
+  trait HasChecker extends HasInput {
+    val checker: Checker
+  }
+
+  trait HasTypechecker extends HasIO {
+    val typechecker: Typechecker
+  }
+
+  // This is the maximal interface, that should always be the greatest upper
+  // bound on all combinations of option groups
+  trait HasAll extends HasTypechecker with HasChecker
+
+  //////////////////
+  // Constructors //
```

Comment:
These constructors is another place where I wish we could get away with a bit less boilerplate, but I don't know how it could be done.